Build continues if go compile fails

## Expected Behavior

If the make action in BINARY_BUILD() fails, the entire build should stop and fail with a non-zero exit code

## Actual Behavior

The build continues

## Steps to Reproduce the Problem

Type some random characters in any of the *.go files and then build a workshop. You will see errors in the log, but the build will continue. 

## Version

=< 3.0

## Additional Information

```
go get -d
env CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -ldflags="-X main.version=1.0.5o" -o ./bin/linux64/tw_provision
# _/build/go/tw_provision_oracle
./tw_provision.go:1326:2: undefined: bam
make: *** [Makefile:12: build] Error 2
echo $GOPATH
/go
go get -d
env CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -ldflags="-X main.version=1.0.7p" -o ./bin/linux64/tw_provision
BATCH 1
```

--- a/demo-workshops/shared/library.sh
+++ b/demo-workshops/shared/library.sh
@@ -229,7 +229,7 @@ function BINARY_BUILD() {
   cd $GODIR
   for each in `ls`; do
     cd ${GODIR}/${each}
-    make build
+    make build || exit 1
   done
 }
 

--- a/demo-workshops/shared/packer_build.sh
+++ b/demo-workshops/shared/packer_build.sh
@@ -35,4 +35,5 @@ function ALL() {
 		PACKER_BUILD ${1}
 	fi
 } 2>&1 | tee ${WORKDIR}/WAIT.log
+[[ ${PIPESTATUS[0]} -ne 0 ]] && ERROR
 READY

--- a/demo-workshops/tcw-oracle/workshop.yaml
+++ b/demo-workshops/tcw-oracle/workshop.yaml
@@ -28,7 +28,7 @@ amis:
 - delphix-tcw-jumpbox-oracle12:
     packer:
       batch: 2
-      md5sum: ef154f8461af8d03bd5738177905dfd5
+      md5sum: 8f1fcd63be518baf52dfe4541bb7291c
       name: delphix-tcw-jumpbox-oracle12.json
 - delphix-tcw-source-oracle12:
     packer:

--- a/demo-workshops/tcw-oracle/workshop.yaml
+++ b/demo-workshops/tcw-oracle/workshop.yaml
@@ -28,7 +28,7 @@ amis:
 - delphix-tcw-jumpbox-oracle12:
     packer:
       batch: 2
-      md5sum: 8f1fcd63be518baf52dfe4541bb7291c
+      md5sum: b1bada36a415b35498700684ca63c41e
       name: delphix-tcw-jumpbox-oracle12.json
 - delphix-tcw-source-oracle12:
     packer:

--- a/demo-workshops/tcw-pg/workshop.yaml
+++ b/demo-workshops/tcw-pg/workshop.yaml
@@ -28,7 +28,7 @@ amis:
 - delphix-tcw-jumpbox-postgres11:
     packer:
       batch: 2
-      md5sum: df34b98c999e70724250a9baa8013cc5
+      md5sum: 14cd109fb7585414f5cf745619edaf72
       name: delphix-tcw-jumpbox-postgres.json
 - delphix-centos7-source-postgres11:
     packer:

--- a/demo-workshops/tcw-pg/workshop.yaml
+++ b/demo-workshops/tcw-pg/workshop.yaml
@@ -28,7 +28,7 @@ amis:
 - delphix-tcw-jumpbox-postgres11:
     packer:
       batch: 2
-      md5sum: 5109929d423e5f72149d202c6ba79cf9
+      md5sum: df34b98c999e70724250a9baa8013cc5
       name: delphix-tcw-jumpbox-postgres.json
 - delphix-centos7-source-postgres11:
     packer:

--- a/demo-workshops/tcw-pg/workshop.yaml
+++ b/demo-workshops/tcw-pg/workshop.yaml
@@ -28,7 +28,7 @@ amis:
 - delphix-tcw-jumpbox-postgres11:
     packer:
       batch: 2
-      md5sum: b687d2e33da29ea00cf8bb213cb8c34c
+      md5sum: 5109929d423e5f72149d202c6ba79cf9
       name: delphix-tcw-jumpbox-postgres.json
 - delphix-centos7-source-postgres11:
     packer:

--- a/go/tw_provision_oracle/Makefile
+++ b/go/tw_provision_oracle/Makefile
@@ -9,10 +9,10 @@ local-debug:
 build:
 	echo $$GOPATH
 	go get -d
-	env CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -ldflags="-X main.version=1.0.5o" -o ./bin/linux64/tw_provision
+	env CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -ldflags="-X main.version=1.0.6o" -o ./bin/linux64/tw_provision
 	
 buildmac:
 	echo $$GOPATH
 	go get -d -v
-	env GOOS=darwin GOARCH=amd64 go build -installsuffix 'static' -ldflags="-X main.version=1.0.5o" -v -o ./bin/macos/tw_provision
+	env GOOS=darwin GOARCH=amd64 go build -installsuffix 'static' -ldflags="-X main.version=1.0.6o" -v -o ./bin/macos/tw_provision
 	

--- a/go/tw_provision_oracle/tw_provision.go
+++ b/go/tw_provision_oracle/tw_provision.go
@@ -243,10 +243,10 @@ func (c *myClient) createDatasetGroup(groupName string, wait bool) (results map[
 // groupNameList: a slice of groupNames to create
 func (c *myClient) batchCreateDatasetGroup(groupNameList ...string) (resultsList []map[string]interface{}, err error) {
 	for _, v := range groupNameList {
-		if action, err := c.createDatasetGroup(v, false); err != nil {
+		if result, err := c.createDatasetGroup(v, false); err != nil {
 			return nil, err
-		} else if action != nil {
-			resultsList = append(resultsList, action)
+		} else if result != nil && (result["job"] != nil || result["action"] != nil) {
+			resultsList = append(resultsList, result)
 		}
 	}
 	c.jobWaiter(resultsList...)
@@ -404,7 +404,7 @@ func (c *myClient) linkMaskingJob() (results map[string]interface{}, err error) 
 	return action, err
 }
 
-func (c *myClient) provisionVDB(vdbParams VDBParams, wait bool) (results map[string]interface{}, err error) {
+func (c *myClient) provisionVDB(vdbParams VDBParams, retPol string, wait bool) (results map[string]interface{}, err error) {
 	namespace := "database"
 	maskingJSON := ""
 	var maskingJobRef interface{}
@@ -486,27 +486,34 @@ func (c *myClient) provisionVDB(vdbParams VDBParams, wait bool) (results map[str
 		if wait {
 			c.jobWaiter(action)
 		}
+		err = c.applyDemoRetentionPolicy(retPol, action["result"].(string))
 		return action, err
 	} else if err != nil {
 		return nil, err
 	} else {
 		log.Debug(vdbObjRef)
 		log.Infof("%s already exists", vdbParams.vdbName)
-		return nil, err
+		results := make(map[string]interface{})
+		results["result"] = vdbObjRef.(string)
+		err = c.applyDemoRetentionPolicy(retPol, vdbObjRef.(string))
+		return results, err
 	}
 }
 
 // batchProvisionVDB takes one parameter:
 // vdbParamsList: a slice of VDBParams to create
-func (c *myClient) batchProvisionVDB(vdbParamsList ...VDBParams) (resultsList []map[string]interface{}, err error) {
+func (c *myClient) batchProvisionVDB(retPol string, vdbParamsList ...VDBParams) (resultsList []map[string]interface{}, err error) {
 	for _, v := range vdbParamsList {
-		if action, err := c.provisionVDB(v, false); action != nil && err == nil {
-			resultsList = append(resultsList, action)
+		if result, err := c.provisionVDB(v, retPol, false); result != nil && err == nil && (result["job"] != nil || result["action"] != nil) {
+			resultsList = append(resultsList, result)
 		} else if err != nil {
 			return nil, err
 		}
 	}
-	err = c.jobWaiter(resultsList...)
+
+	if resultsList != nil {
+		c.jobWaiter(resultsList...)
+	}
 
 	return resultsList, err
 }
@@ -697,13 +704,15 @@ func (c *myClient) populateMasking() (err error) {
 // containerList: a slice of PatientContainer to create
 func (c *myClient) batchCreateSelfServiceContainer(containerList ...PatientContainer) (resultsList []map[string]interface{}, err error) {
 	for _, v := range containerList {
-		if action, err := c.createSelfServiceContainer(v, false); action != nil && err == nil {
-			resultsList = append(resultsList, action)
+		if result, err := c.createSelfServiceContainer(v, false); result != nil && err == nil && (result["job"] != nil || result["action"] != nil) {
+			resultsList = append(resultsList, result)
 		} else if err != nil {
 			return nil, err
 		}
 	}
-	c.jobWaiter(resultsList...)
+	if resultsList != nil {
+		c.jobWaiter(resultsList...)
+	}
 
 	return resultsList, err
 }
@@ -1173,7 +1182,7 @@ func main() {
 		maskingJob: "Patients Mask",
 	}
 
-	_, err = virtualizationClient.provisionVDB(patmm, true)
+	_, err = virtualizationClient.provisionVDB(patmm, retPolicy, true)
 	if err != nil {
 		logger.Fatal(err)
 	}
@@ -1250,7 +1259,7 @@ func main() {
 		envName:   "devdb",
 	}
 
-	_, err = virtualizationClient.batchProvisionVDB(devdb, testdb, prep1, prep2, prep3, prep4, prep5, prep6)
+	_, err = virtualizationClient.batchProvisionVDB(retPolicy, devdb, testdb, prep1, prep2, prep3, prep4, prep5, prep6)
 	if err != nil {
 		logger.Fatal(err)
 	}

--- a/go/tw_provision_oracle/tw_provision.go
+++ b/go/tw_provision_oracle/tw_provision.go
@@ -1015,6 +1015,49 @@ func (c *myClient) updateMaskingService() (map[string]interface{}, error) {
 	return result, err
 }
 
+func (c *myClient) createDemoRetentionPolicy() (string, error) {
+	namespace := "policy"
+	policyName := "Demo"
+	if policyRef, err := c.findObjectByNameReturnReference(namespace, policyName); policyRef == nil && err == nil {
+		postBody := fmt.Sprintf(`{
+		"dataDuration": 2,
+		"dataUnit": "YEAR",
+		"logDuration": 1,
+		"logUnit": "YEAR",
+		"customized": false,
+		"name": "%s",
+		"type": "RetentionPolicy"
+		}`, policyName)
+
+		action, _, err := c.httpPost(namespace, postBody)
+		if err != nil {
+			return "", err
+		}
+
+		c.jobWaiter(action)
+		return action["result"].(string), err
+	} else {
+		return policyRef.(string), err
+	}
+}
+
+func (c *myClient) applyDemoRetentionPolicy(policyRef, containerRef string) error {
+	namespace := "policy"
+	postBody := fmt.Sprintf(`{
+			"target":"%s",
+			"type":"PolicyApplyTargetParameters"
+		}`, containerRef)
+
+	action, _, err := c.httpPost(fmt.Sprintf("%s/%s/apply", namespace, policyRef), postBody)
+	if err != nil {
+		return err
+	}
+
+	c.jobWaiter(action)
+	return err
+
+}
+
 var (
 	opts             Options
 	parser           = flags.NewParser(&opts, flags.Default)
@@ -1042,6 +1085,11 @@ func main() {
 	maskingCR = NewClientRequest("admin", "Admin-12", fmt.Sprintf("http://%s/masking/api", opts.DDPMaskName))
 	maskingClient = maskingCR.initResty()
 	maskingClient.setInitialMaskingPassword()
+
+	retPolicy, err := virtualizationClient.createDemoRetentionPolicy()
+	if err != nil {
+		log.Fatal(err)
+	}
 
 	result, err := virtualizationClient.updateMaskingService()
 	if err != nil {
@@ -1084,6 +1132,11 @@ func main() {
 		logger.Fatal(err)
 	}
 	_, err = virtualizationClient.linkDatabase(true)
+	if err != nil {
+		logger.Fatal(err)
+	}
+
+	err = virtualizationClient.applyDemoRetentionPolicy(retPolicy, dSourceRef["result"].(string))
 	if err != nil {
 		logger.Fatal(err)
 	}

--- a/go/tw_provision_oracle/tw_provision.go
+++ b/go/tw_provision_oracle/tw_provision.go
@@ -333,7 +333,9 @@ func (c *myClient) linkDatabase(wait bool) (results map[string]interface{}, err 
 	} else {
 		log.Debug(dbObjRef)
 		log.Infof("%s already exists", dbName)
-		return nil, err
+		results := make(map[string]interface{})
+		results["result"] = dbObjRef.(string)
+		return results, err
 	}
 }
 
@@ -1131,7 +1133,8 @@ func main() {
 	if err != nil {
 		logger.Fatal(err)
 	}
-	_, err = virtualizationClient.linkDatabase(true)
+
+	dSourceRef, err := virtualizationClient.linkDatabase(true)
 	if err != nil {
 		logger.Fatal(err)
 	}

--- a/go/tw_provision_postgres/Makefile
+++ b/go/tw_provision_postgres/Makefile
@@ -10,10 +10,10 @@ local-debug:
 build:
 	echo $$GOPATH
 	go get -d
-	env CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -ldflags="-X main.version=1.0.7p" -o ./bin/linux64/tw_provision
+	env CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -ldflags="-X main.version=1.0.8p" -o ./bin/linux64/tw_provision
 	
 buildmac:
 	echo $$GOPATH
 	go get -d -v
-	env GOOS=darwin GOARCH=amd64 go build -installsuffix 'static' -ldflags="-X main.version=1.0.7p" -v -o ./bin/macos/tw_provision
+	env GOOS=darwin GOARCH=amd64 go build -installsuffix 'static' -ldflags="-X main.version=1.0.8p" -v -o ./bin/macos/tw_provision
 	

--- a/go/tw_provision_postgres/tw_provision.go
+++ b/go/tw_provision_postgres/tw_provision.go
@@ -704,7 +704,7 @@ func (c *myClient) provisionVDB(vdbParams PGVDBParams, retPol string, wait bool)
 		if wait {
 			c.jobWaiter(action)
 		}
-		err = c.applyDemoRetentionPolicy(retPol, vdbObjRef.(string))
+		err = c.applyDemoRetentionPolicy(retPol, action["result"].(string))
 		return action, err
 	} else if err != nil {
 		return nil, err

--- a/go/tw_provision_postgres/tw_provision.go
+++ b/go/tw_provision_postgres/tw_provision.go
@@ -1593,6 +1593,9 @@ func main() {
 	}
 	logger.Info(dSourceRef)
 	err = virtualizationClient.applyDemoRetentionPolicy(retPolicy, dSourceRef["result"].(string))
+	if err != nil {
+		logger.Fatal(err)
+	}
 	err = maskingClient.populateMasking(proddb.dbUser, proddb.dbPass, patmm.port)
 	if err != nil {
 		logger.Fatal(err)

--- a/go/tw_provision_postgres/tw_provision.go
+++ b/go/tw_provision_postgres/tw_provision.go
@@ -277,10 +277,10 @@ func (c *myClient) createDatasetGroup(groupName string, wait bool) (results map[
 // groupNameList: a slice of groupNames to create
 func (c *myClient) batchCreateDatasetGroup(groupNameList ...string) (resultsList []map[string]interface{}, err error) {
 	for _, v := range groupNameList {
-		if action, err := c.createDatasetGroup(v, false); err != nil {
+		if result, err := c.createDatasetGroup(v, false); err != nil {
 			return nil, err
-		} else if action != nil {
-			resultsList = append(resultsList, action)
+		} else if result != nil && (result["job"] != nil || result["action"] != nil) {
+			resultsList = append(resultsList, result)
 		}
 	}
 	c.jobWaiter(resultsList...)
@@ -604,7 +604,7 @@ func (c *myClient) provisionPVDB(vdbParams PVDBParams, wait bool) (results map[s
 }
 
 // provisionVDB provisions VDBs
-func (c *myClient) provisionVDB(vdbParams PGVDBParams, wait bool) (results map[string]interface{}, err error) {
+func (c *myClient) provisionVDB(vdbParams PGVDBParams, retPol string, wait bool) (results map[string]interface{}, err error) {
 	namespace := "database"
 
 	if vdbObjRef, err := c.findObjectByNameReturnReference(namespace, vdbParams.vdbName); vdbObjRef == nil && err == nil {
@@ -704,13 +704,17 @@ func (c *myClient) provisionVDB(vdbParams PGVDBParams, wait bool) (results map[s
 		if wait {
 			c.jobWaiter(action)
 		}
+		err = c.applyDemoRetentionPolicy(retPol, vdbObjRef.(string))
 		return action, err
 	} else if err != nil {
 		return nil, err
 	} else {
 		log.Debug(vdbObjRef)
 		log.Infof("%s already exists", vdbParams.vdbName)
-		return nil, err
+		results := make(map[string]interface{})
+		results["result"] = vdbObjRef.(string)
+		err = c.applyDemoRetentionPolicy(retPol, vdbObjRef.(string))
+		return results, err
 	}
 }
 
@@ -756,10 +760,10 @@ func (c *myClient) createSourceConfig(scParams AppDataStagedSourceConfig, wait b
 
 // batchProvisionVDB takes one parameter:
 // vdbParamsList: a slice of VDBParams to create
-func (c *myClient) batchProvisionVDB(vdbParamsList ...PGVDBParams) (resultsList []map[string]interface{}, err error) {
+func (c *myClient) batchProvisionVDB(retPol string, vdbParamsList ...PGVDBParams) (resultsList []map[string]interface{}, err error) {
 	for _, v := range vdbParamsList {
-		if action, err := c.provisionVDB(v, false); action != nil && err == nil {
-			resultsList = append(resultsList, action)
+		if result, err := c.provisionVDB(v, retPol, false); result != nil && err == nil && (result["job"] != nil || result["action"] != nil) {
+			resultsList = append(resultsList, result)
 		} else if err != nil {
 			return nil, err
 		}
@@ -964,8 +968,8 @@ func (c *myClient) populateMasking(user, pass string, port int) (err error) {
 // containerList: a slice of PatientContainer to create
 func (c *myClient) batchCreateSelfServiceContainer(containerList ...PatientContainer) (resultsList []map[string]interface{}, err error) {
 	for _, v := range containerList {
-		if action, err := c.createSelfServiceContainer(v, false); action != nil && err == nil {
-			resultsList = append(resultsList, action)
+		if result, err := c.createSelfServiceContainer(v, false); result != nil && err == nil && (result["job"] != nil || result["action"] != nil) {
+			resultsList = append(resultsList, result)
 		} else if err != nil {
 			return nil, err
 		}
@@ -1615,7 +1619,7 @@ func main() {
 	// 	logger.Fatal(err)
 	// }
 
-	_, err = virtualizationClient.provisionVDB(patmm, true)
+	_, err = virtualizationClient.provisionVDB(patmm, retPolicy, true)
 	if err != nil {
 		logger.Fatal(err)
 	}
@@ -1626,7 +1630,7 @@ func main() {
 		logger.Fatal(err)
 	}
 
-	_, err = virtualizationClient.batchProvisionVDB(devdb, testdb, prep1, prep2, prep3, prep4, prep5, prep6)
+	_, err = virtualizationClient.batchProvisionVDB(retPolicy, devdb, testdb, prep1, prep2, prep3, prep4, prep5, prep6)
 	if err != nil {
 		logger.Fatal(err)
 	}

--- a/packer-templates/delphix-tcw-jumpbox-oracle12.json
+++ b/packer-templates/delphix-tcw-jumpbox-oracle12.json
@@ -130,7 +130,7 @@
     "md5sum": "",
     "region": "{{env `AWS_REGION`}}",
     "release": "{{env `RELEASE`}}",
-    "seed": "49a6d5d",
+    "seed": "ca5f2ed",
     "source": "delphix-ubuntu-bionic-guacamole.json",
     "subnet_id": "{{env `AWS_SUBNET_ID`}}",
     "vpc_id": "{{env `AWS_VPC_ID`}}"

--- a/packer-templates/delphix-tcw-jumpbox-oracle12.json
+++ b/packer-templates/delphix-tcw-jumpbox-oracle12.json
@@ -130,7 +130,7 @@
     "md5sum": "",
     "region": "{{env `AWS_REGION`}}",
     "release": "{{env `RELEASE`}}",
-    "seed": "ca5f2ed",
+    "seed": "fba3617",
     "source": "delphix-ubuntu-bionic-guacamole.json",
     "subnet_id": "{{env `AWS_SUBNET_ID`}}",
     "vpc_id": "{{env `AWS_VPC_ID`}}"

--- a/packer-templates/delphix-tcw-jumpbox-postgres.json
+++ b/packer-templates/delphix-tcw-jumpbox-postgres.json
@@ -131,7 +131,7 @@
     "pg_ver": "{{env `PG_VER`}}",
     "region": "{{env `AWS_REGION`}}",
     "release": "{{env `RELEASE`}}",
-    "seed": "49a6d5d",
+    "seed": "6c0c0aa",
     "source": "delphix-ubuntu-bionic-guacamole.json",
     "subnet_id": "{{env `AWS_SUBNET_ID`}}",
     "vpc_id": "{{env `AWS_VPC_ID`}}"

--- a/packer-templates/delphix-tcw-jumpbox-postgres.json
+++ b/packer-templates/delphix-tcw-jumpbox-postgres.json
@@ -131,7 +131,7 @@
     "pg_ver": "{{env `PG_VER`}}",
     "region": "{{env `AWS_REGION`}}",
     "release": "{{env `RELEASE`}}",
-    "seed": "6c0c0aa",
+    "seed": "fba3617",
     "source": "delphix-ubuntu-bionic-guacamole.json",
     "subnet_id": "{{env `AWS_SUBNET_ID`}}",
     "vpc_id": "{{env `AWS_VPC_ID`}}"

--- a/packer-templates/delphix-tcw-jumpbox-postgres.json
+++ b/packer-templates/delphix-tcw-jumpbox-postgres.json
@@ -131,7 +131,7 @@
     "pg_ver": "{{env `PG_VER`}}",
     "region": "{{env `AWS_REGION`}}",
     "release": "{{env `RELEASE`}}",
-    "seed": "fba3617",
+    "seed": "b1af73f",
     "source": "delphix-ubuntu-bionic-guacamole.json",
     "subnet_id": "{{env `AWS_SUBNET_ID`}}",
     "vpc_id": "{{env `AWS_VPC_ID`}}"

